rfcs/0003-mutations.md

```diff
@@ -0,0 +1,619 @@
+
+      // Create ethereum transactions...
+      // Fetch & upload to ipfs...
+
+      // Dispatch a state event through the state updater
+      const { state } = context.graph
+      await state.dispatch("PROGRESS_UPDATE", { progress: 0.5 })
+
+      // Dispatch a custom ext event
+      await state.dispatch("MY_EVENT", { myValue: "..." })
+
+      // Get a copy of the current state
+      const currentState = state.current
+
+      // Send another query using the same client.
+      // This query would result in the graph-node's
+      // entity store being fetched from. You could also
+      // execute another mutation here if desired.
+      const { client } = context
+      await client.query({
+        query: gql`
+          myEntity (id: "${id}") {
+            id
+            name
+            value
+          }
+        }`
+      })
+
+      // Fetch datasource addresses & abis from the subgraph manifest
+      const { MyContract } = context.graph.datasources
+      await MyContract.abi
+      await MyContract.address
+
+      ...
+    },
+    async setEntityName (source: any, args: any, context: any) {
+      ...
+    }
+  }
+}
+
+/// Config Generators
+const config: ConfigGenerators = {
+  // These function arguments are passed in by the dApp
+  ethereum (provider: AsyncSendable): Web3Provider {
+    return new ethers.providers.Web3Provider(provider)
+  },
+  ipfs (provider: string): IPFS {
+    return new IPFS(provider)
+  },
+  // Example of a custom config property
+  property: {
+    // Generators can be nested
+    a: (value: string) => { },
+    b: (value: string) => { }
+  }
+}
+
+/// (optional) Ext State, Events, and State Builder
+
+// Ext State
+interface State {
+  myValue: string
+}
+
+// Ext Events
+interface MyEvent extends EventPayload {
+  myValue: string
+}
+
+interface EventMap extends EventTypeMap {
+  "MY_EVENT": MyEvent
+}
+
+// Ext State Builder
+const stateBuilder: StateBuilder<State, EventMap> = {
+  getInitialState(): State {
+    return {
+      myValue: ""
+    }
+  },
+  reducers: {
+    "MY_EVENT": (state: MutationState<State>, payload: MyEvent) => {
+      return {
+        myValue: payload.myValue
+      }
+    },
+    "PROGRESS_UPDATE": (state: MutationState<State>, payload: ProgressUpdateEvent) => {
+      // Do something custom...
+    }
+  },
+  // Catch-all reducer...
+  reducer: (state: MutationState<State>, event: Event) => {
+    switch (event.name) {
+      case "TRANSACTION_CREATED":
+        // Do something custom...
+        break;
+    }
+  }
+}
+
+export default {
+  resolvers,
+  config,
+  stateBuilder
+}
+
+export {
+  State,
+  EventMap,
+  MyEvent
+}
+```
+
+NOTE: If the resolvers module exports a `stateBuilder`, it's expected that the mutations manifest has a `resolvers.types` file defined. The following types are expected to be defined in the .d.ts type definition file:
+  - `State`
+  - `EventMap`
+  - Any `EventPayload` interfaces defined within the `EventMap`
```

Comment:
My personal opinion is that the developer should still be expected to write the interfaces manually within the .d.ts file. This way dApps will know what properties are within the state.

Deferring to you to make the final call here.

Reply:
@Jannis I've changed this to state that it is required to have a resolvers.types file. Additionally I've added the `Config` type to list of require exported types.

This is because, when executing a mutation without Apollo, the developer needs to define a context for the execution to use. This `MutationContext` type requires type parameters for `Config`, `State`, and `EventMap`. An example of this can be seen here: https://github.com/dOrgTech/the-graph-mutations-spec/blob/586018191737a7606957c49cecd081021ac7abd1/packages/mutations/example/dapp.ts#L29-L52

Reply:
Ok, let's leave it as mandatory for now then.